File: server/services/opentok.js
/* Config */
const config = require('../config/credentials');
const apiKey = config.opentok.key;
const apiSecret = config.opentok.secret;

const OpenTok = require('opentok');
const Promise = require('bluebird');
const R = require('ramda');
const OT = Promise.promisifyAll(new OpenTok(apiKey, apiSecret));

/** Private */

const defaultSessionOptions = { mediaMode: 'routed' };

/**
 * Returns options for token creation based on user type
 * @param {String} userType Host, guest, or viewer
 */
const tokenOptions = user => {
  const role = {
    instructor: 'moderator',
    student: 'publisher',
    auditor: 'subscriber',
  }[user.role];

  const data = JSON.stringify({
    id: user.id,
    role: user.role,
    name: user.name
  });

  return { role, data };
};

/** Exports */

/**
 * Create an OpenTok token
 * @param {String} sessionId
 * @param {String} userType instructor, student, auditor
 * @returns {Object} Credentials, including apiKey, sessionId, and token
 */
const createToken = (sessionId, user) => {
  const token = OT.generateToken(sessionId, tokenOptions(user));
  return { apiKey, sessionId, token };
};


/**
 * Create an OpenTok session
 * @param {Object} [options]
 * @returns {Promise} <Resolve => {Object}, Reject => {Error}>
 */
const createSession = options =>
  new Promise((resolve, reject) => {
    OT.createSession(R.defaultTo(defaultSessionOptions)(options), (err, session) => {
      if (err) { reject(err); }
      resolve(session);
    });
  });

module.exports = {
  createSession,
  createToken
};

File: scala-server/src/main/scala/OpenTok.scala
import java.io.InputStream
import scala.io.Source
import com.opentok.OpenTok
import com.opentok.Session
import com.opentok.SessionProperties
import com.opentok.MediaMode
import com.opentok.TokenOptions;
import com.opentok.Role;
import io.circe._, io.circe.parser._, io.circe.generic.semiauto._, io.circe.parser.decode, io.circe.syntax._
import cats.syntax.either._


case class OpentokCredentials(apiKey: Int, apiSecret: String)

object Opentok {
  /** Initialize */
  val credentialsJSON: String = {
    val stream: InputStream = getClass.getResourceAsStream("/opentokCredentials.json")
    scala.io.Source.fromInputStream(stream).mkString
  }
  implicit val credentialsDecoder: Decoder[OpentokCredentials] = deriveDecoder[OpentokCredentials]
  val credentials = decode[OpentokCredentials](credentialsJSON).getOrElse(null);

 val opentok: OpenTok = new OpenTok(credentials.apiKey, credentials.apiSecret)

  def createSession = {
    val session: Session = opentok.createSession(new SessionProperties.Builder()
      .mediaMode(MediaMode.ROUTED)
      .build());
    session
  }

  def createToken(sessionId: String, user: User) = {

    val tokenRoles: Map[String, Role] = Map(
      "instructor" -> Role.MODERATOR,
      "student" -> Role.PUBLISHER,
      "auditor" -> Role.SUBSCRIBER
    )
    val tokenData: Json = Map(
      "id" -> user.id,
      "role" -> user.role.get,
      "name" -> user.name
    ).asJson

    val token = opentok.generateToken(sessionId, new TokenOptions.Builder()
      .role(tokenRoles(user.role.get))
      .data("name=Johnny")
      .build());
  }
}






